### PR TITLE
Update DNQ fixes issue #4418

### PR DESCRIPTION
**Pull request by RadialFunc**
_Place an "x" in the corresponding checkbox if it is done or does not apply to this pull request._

- [x ] I have tested my code.
- [x ] I have added my solution according to the steps [here](https://usaco.guide/general/adding-solution#steps).
- [x ] I have followed the code conventions mentioned [here](https://usaco.guide/general/adding-solution/#code-conventions).
  - I understand that if it is clear that I have not attempted to follow these conventions, my PR will be closed.
  - If changes are requested, I will re-request a review after addressing them.
- [x ] I have linked this PR to any issues that it closes.

### CODE REVIEW

**Comment by SansPapyrus683** [comment]
is this for a bounty

**Comment by RadialFunc** [comment]
it is

**Comment by SansPapyrus683** [comment]
image

this part doesn't seem addressed

**Comment by RadialFunc** [comment]
ok i think i added the code

File: content/5_Plat/DC-DP.mdx [hunk 1]
```diff
@@ -56,9 +56,73 @@ Let $dp(i,k)$ denote the location of the last door if we place $k$ doors optimal
...
-
+Here is an implementation of the above algorithm.
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+#define int long long
+#define V vector
+#define vi V<int>
+
+const int INF = 1e18;
```
Comment (SansPapyrus683) [inline content/5_Plat/DC-DP.mdx:67]: use INT32_MAX

File: content/5_Plat/DC-DP.mdx [hunk 2]
```diff
@@ -56,9 +56,73 @@ Let $dp(i,k)$ denote the location of the last door if we place $k$ doors optimal
...
-[official editorial](http://www.usaco.org/current/data/sol_cbarn_platinum_feb16.html) for an implementation.
-
+Here is an implementation of the above algorithm.
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+#define int long long
+#define V vector
+#define vi V<int>
+
```
Comment (SansPapyrus683) [inline content/5_Plat/DC-DP.mdx:66]: image

File: content/5_Plat/DC-DP.mdx [hunk 3]
```diff
@@ -56,9 +56,73 @@ Let $dp(i,k)$ denote the location of the last door if we place $k$ doors optimal
...
+const int INF = 1e18;
+
+// most array sizes will be doubled to prevent out of bounds
+
+V<vi> calc(2000, vi(1001,0)); // calc[i][j] stores the # of steps to get all cows distance j away to door i
+V<vi> dp(8, vi(1001,INF));
```
Comment (SansPapyrus683) [inline content/5_Plat/DC-DP.mdx]: magic numbers

File: content/5_Plat/DC-DP.mdx [hunk 4]
```diff
@@ -56,9 +56,73 @@ Let $dp(i,k)$ denote the location of the last door if we place $k$ doors optimal
...
+		}else if(cost==bestcost && lst < last){
+			last = lst;
+		}
+	}
+	dp[k][mid] = bestcost;
+	if(begin==end) return;
+	compdp(k, begin, mid, rl, last);
+	compdp(k, mid+1, end, last, rr);
+}
+signed main(){
```
Comment (SansPapyrus683) [inline content/5_Plat/DC-DP.mdx]:
```suggestion

int main() {
```

File: content/5_Plat/DC-DP.mdx [hunk 5]
```diff
@@ -56,9 +56,73 @@ Let $dp(i,k)$ denote the location of the last door if we place $k$ doors optimal
...
+	int n,k;
+	cin >> n >> k;
+	vi a(2*n); 
+	for(int i=0; i<n; i++){
+		cin >> a[i];
+		a[i+n] = a[i];
```
Comment (SansPapyrus683) [inline content/5_Plat/DC-DP.mdx]: spaces between operators

File: content/5_Plat/DC-DP.mdx [hunk 6]
```diff
@@ -56,9 +56,73 @@ Let $dp(i,k)$ denote the location of the last door if we place $k$ doors optimal
...
+	// dp[i][j] stores the answer for doors 0,1,.., j and i doors open
+	
+	// rotate stores where we start in the linear representation
+	int ans = INF;
+	for(rot=0; rot<n; rot++){
+		for(int i=0; i<n; i++){
+			dp[0][i] = calc[rot][i+1];
+		}
+		for(int i=1; i<k; i++){
+			for(int j=0; j<n; j++) dp[i][j] = INF;
```
Comment (SansPapyrus683) [inline content/5_Plat/DC-DP.mdx]: braces

**Review by SansPapyrus683 — CHANGES_REQUESTED** [review]
have you seen my standards before?

**Comment by RadialFunc** [comment]
i fixed everything

File: content/5_Plat/DC-DP.mdx [hunk 7]
```diff
@@ -32,15 +50,84 @@ Allows you to reduce $\mathcal{O}(N^2)$ to $\mathcal{O}(N\log N)$.
...
+Since the monotonicity condition is held, we can now perform Divide & Conquer DP. Fix the value of $k$ and compute $dp(n/2, k)$. Then compute it for the left and right halves of the array. This leads to a runtime of $O(nk \log n)$. Since we need to check $n$ rooms for the optimal first barn position, the overall runtime is $O(n^2 k \log n)$.
+
+Here is an implementation of the above algorithm.
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+#define ll long long
+const ll INF = 1e18;
+
+// most array sizes will be doubled to prevent out of bounds
```
Comment (SansPapyrus683) [inline content/5_Plat/DC-DP.mdx:66]: uuuhhh what

File: content/5_Plat/DC-DP.mdx [hunk 8]
```diff
@@ -32,15 +50,84 @@ Allows you to reduce $\mathcal{O}(N^2)$ to $\mathcal{O}(N\log N)$.
...
-[official editorial](http://www.usaco.org/current/data/sol_cbarn_platinum_feb16.html).
-
+We iterate through the possibilities of the location of the first door. For each of the first doors, we can now view the barn linearly. All further calculations will be done assuming the barn is a linear sequence of doors starting at the first opened door.
+
+Let $dp(i,k)$ denote the location of the last door if we place $k$ doors optimally among the first $i$ rooms. The idea is that $dp(i,k) \leq dp(i+1, k)$. Assume this was not true for the sake of contradiction. Then $dp(i,k) > dp(i+1,k)$, so we also have $dp(i+1,k) \leq i$. But then we could have used the best possible setup for $(i+1,k)$ in the $(i,k)$ setup as well, since all open doors are among the first $i$ rooms anyways.
+
+Since the monotonicity condition is held, we can now perform Divide & Conquer DP. Fix the value of $k$ and compute $dp(n/2, k)$. Then compute it for the left and right halves of the array. This leads to a runtime of $O(nk \log n)$. Since we need to check $n$ rooms for the optimal first barn position, the overall runtime is $O(n^2 k \log n)$.
+
+Here is an implementation of the above algorithm.
+```cpp
```
Comment (SansPapyrus683) [inline content/5_Plat/DC-DP.mdx:60]: oh yeah use language section

File: content/5_Plat/DC-DP.mdx (hunk 1 above)
Comment (SansPapyrus683) [reply content/5_Plat/DC-DP.mdx]: oops

File: content/5_Plat/DC-DP.mdx [hunk 9]
```diff
@@ -32,15 +52,104 @@ Allows you to reduce $\mathcal{O}(N^2)$ to $\mathcal{O}(N\log N)$.
...
+
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+#define ll long long
+const ll INF = 1e18;
+
+// to make implementing a cyclic array easier, we double the size
+const int maxn = 1000;
+const int maxk = 7;
```
Comment (SansPapyrus683) [inline content/5_Plat/DC-DP.mdx:86]: rename to `MAX_N`a dn `MAX_K` respectively

File: content/5_Plat/DC-DP.mdx [hunk 10]
```diff
@@ -32,15 +52,104 @@ Allows you to reduce $\mathcal{O}(N^2)$ to $\mathcal{O}(N\log N)$.
...
+		a[i + n] = a[i];
+	}
+
+	for (int i = 0; i < n; i++)
+		for (int j = 1; j <= n; j++) {
+			calc[i][j] = calc[i + n][j] =
+			    calc[i][j - 1] + (ll)a[i + j - 1] * (j - 1);
+		}
+
+	// dp[i][j] stores the answer for doors 0,1,.., j and i doors open
```
Comment (SansPapyrus683) [inline content/5_Plat/DC-DP.mdx:133]: put this at the top of `dp`?

File: content/5_Plat/DC-DP.mdx [hunk 11]
```diff
@@ -32,15 +52,104 @@ Allows you to reduce $\mathcal{O}(N^2)$ to $\mathcal{O}(N\log N)$.
...
+		ll cost = dp[k - 1][lst - 1] + calc[lst + rot][mid - lst + 1];
+		if (cost < bestcost) {
+			bestcost = cost;
+			last = lst;
+		} else if (cost == bestcost && lst < last) {
+			last = lst;
+		}
+	}
+	dp[k][mid] = bestcost;
+	if (begin == end) return;
```
Comment (SansPapyrus683) [inline content/5_Plat/DC-DP.mdx:111]: braces

File: content/5_Plat/DC-DP.mdx [hunk 12]
```diff
@@ -32,15 +52,104 @@ Allows you to reduce $\mathcal{O}(N^2)$ to $\mathcal{O}(N\log N)$.
...
+vector<vector<ll>> calc(2 * maxn, vector<ll>(maxn + 1, 0));
+vector<vector<ll>> dp(maxk + 1, vector<ll>(maxn + 1, INF));
+int rot;
+
+void compdp(int k, int begin, int end, int rl, int rr) {
+	// fixed k, begin and end are the ends of the array, rl and rr are the
+	// bounds on the last door used
+	int mid = (begin + end) / 2;
+
+	ll bestcost = INF;
```
Comment (SansPapyrus683) [inline content/5_Plat/DC-DP.mdx:98]: just `best` is fine

**Review by SansPapyrus683 — CHANGES_REQUESTED** [review]
i'd say we're 70% there

File: content/5_Plat/DC-DP.mdx (hunk 1 above)
Comment (SansPapyrus683) [reply content/5_Plat/DC-DP.mdx]: don't think this was addressed?

**Review by SansPapyrus683 — APPROVED** [review]
send ss of updated code AC and we're good

**Review by ryanchou-dev — APPROVED** [review]
good 👍

**Pull request merged**